src/models/StyledComponent.js

```diff
@@ -117,52 +169,72 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
     }
 
     componentWillUnmount() {
-      if (this.unsubscribe) {
-        this.unsubscribe()
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        return
       }
+
+      this.unsubscribeFromContext()
     }
 
-    render() {
-      const { innerRef } = this.props
+    makeClassName(): string {
+      // I know this is ugly, but I have a really surprigins JSPerf
+      // pen showing this is ~11x faster than running [...classes].filter(Boolean).join(' ')
+      // (and uses way less memory from the two array allocations in ^)
+      // and  ~ 36 % faster than always adding a whitespace and then calling .trim at the end
+      const { className: propsClassName } = this.props
+      const { className: attrsClassName } = this.attrs
       const { generatedClassName } = this.state
-      const { styledComponentId, target } = this.constructor
+      const { styledComponentId } = this.constructor
 
-      const isTargetTag = isTag(target)
+      let className = propsClassName || ''
+      if (styledComponentId) {
```

Comment:
Sure! added some more tests and published: https://jsperf.com/testing-classname-building-performance. My math was wrong from rounding, but still a large improvement. I'll add it in a comment once this is separated to its own patch.